test/test_common/utility.h

```diff
@@ -80,6 +81,16 @@ namespace Envoy {
   EXPECT_THAT_THROWS_MESSAGE(statement, expected_exception,                                        \
                              ::testing::Not(::testing::ContainsRegex(regex_str)))
 
+// Expect that the statement hits an ENVOY_BUG containing the specified message.
+#ifdef NDEBUG
+// ENVOY_BUGs in release mode log fatal.
+#define EXPECT_ENVOY_BUG(statement, message) EXPECT_LOG_CONTAINS("error", message, statement)
```

Comment:
Does ENVOY_BUG implement any form of exponential backoff?  If yes, this macro may not catch all occurances, would result in tests that are order dependent.

Does ENVOY_BUG at least reset counters between tests?

Reply:
ping

Reply:
ENVOY_BUG will log always, with the expectation that this should not be a performance issue right now
https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/common/common/assert.cc#L56

Stats is implemented with exponential backoff, but they are rooted at the server

Reply:
FWIW I think logging should also be exponential backoff. For example, if we hit the ENVOY_BUG in the recent buffer PR we would be in pretty bad shape I think. cc @ggreenway

Reply:
`// ENVOY_BUG logging and actions are invoked only on power-of-two instances per log line.`

Reply:
Is the documentation incorrect?

Reply:
Yes, it was fixed in this commit: https://github.com/envoyproxy/envoy/pull/14800/commits/4a36040090def9f5cdb0b72f3a8549d743687760

Reply:
Thanks @ggreenway for the reference, I checked the implementation and both logging and action done on power-of-two:
https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/common/common/assert.h#L183-L184.

It looks like the comments I referenced earlier was inconsistent. I'll fix that and make it so that counters are reset in tests.

Reply:
Done! I added a test for this (so that I could make sure it worked. Let me know if I should remove the test of tests).

`assert_test` and `assert2_test` both call into a common ENVOY_BUG and verify that they hit the two consecutive power of twos (1 and 2)